### DISCUSSION TITLE
Embedded mode for cpp client?

### DISCUSSION THREAD

**Discussion by mlawsonca** [hazelcast/hazelcast-cpp-client#931] [Q&A]
Just to confirm, is it correct that the cpp client cannot be used in embedded mode and instead must be run in client/server mode? If it can be embedded, can you point me towards any resources on how to configure it to run in embedded mode with the cpp client?

**Comment by ihsandemir** [comment]
Hello @mlawsonca Thanks for the question. Currently, the C++ client library works as client/server mode only. Hence, you need to [start the server](https://docs.hazelcast.com/imdg/latest/getting-started.html) in a separate process ([additional doc](https://github.com/hazelcast/hazelcast-cpp-client/blob/master/Reference_Manual.md#12-starting-hazelcast-imdg-cluster)). 

Can you please describe your use case and the need why you want to use it as embedded? What is your environment and constraints? May be we can help more with these extra information.

**Comment by mlawsonca** [comment]
My use case: multiple nodes running in parallel with database per node. This allows for parallel data management without any coordination between the nodes. Installing the server and running it in the background could work (so long as it doesn't require pinging the server, which is disabled on this machine). Embedding the database just removes this extra complication, and makes it easier to manage the amount of resources dedicated to the database.

**Comment by ihsandemir** [comment]
@mlawsonca What did you mean by `so long as it doesn't require pinging the server, which is disabled on this machine`? Do you mean the ICMP ping messages? Client does not use ICMP messages, it always talks to the server at a certain tcp port via TCP/IP, this port is usually for example 5701. 

By the way, we even have [near-cache feature](https://github.com/hazelcast/hazelcast-cpp-client/blob/master/Reference_Manual.md#782-near-cache) at the client side which makes frequently accessed data being accessed faster just from local cache.

A lot of deployments actually prefer client/server deployments compared to embedded usage so that the server and the client application are in different processes which also prevents any shared resource problems(memory, threads, etc.) between user application and the server. Of course, this really depends on the nature of your application. On some use cases embedded usage may be preferred as well, i really recommend the client/server unless there is a good reason for embedded usage.

**Comment by mlawsonca** [comment]
Yes, I mean ICMP ping messages. I found a few databases that first ping the server before connecting using TCP/IP. 

And, yes, I have seen the near-caches. That is a very useful feature. 

Thanks for your help!
